### ISSUE TITLE
[utilities] keyboard navigation for hidden elements

### ISSUE THREAD

**Issue by Alchez** [agritheory/stonecrop#15]
**Problem:** Handling keyboard navigation for hidden elements (e.g. collapsible tree-views)

**Context:** There is [no single perfect](https://stackoverflow.com/a/59096915) way to detect hidden elements in a DOM tree using Javascript

**Possible Solution(s):**

- We implement a combination of the techniques in the above link to filter out hidden elements in the composable, and not attach the listener to those.
  - **Pro:** This will not break keyboard navigation, since hidden elements are skipped and therefore focus events will not fail.
  - **Con:** This is implicit behaviour, since the hidden selectors will not receive the event listener. This can be made explicit with a key in the config object.
- We defer the responsibility to the user to filter out the hidden elements before passing it to the composable.
  - **Pro:** The composable is lighter and more reliable, without having to deal with another function and possible edge cases.
  - **Con:** If this expectation is not clear and the user passes hidden elements, that can break keyboard navigation, and that may be bad UX for someone using the composable. This can be avoided with API documentation and/or a warning/error/linter support.

**Comment by agritheory** [comment]
A cross comparison with other common table-navigation idioms shows that Excel, Sheets and Numbers (icloud) all navigate a closed group of rows as if it were one row, skipping all hidden rows inside the grouping. When the group is expanded/ visible, the cells are traversed as per usual.

So the desired (industry standard) behavior is clear, but the implementation is not. I would be comfortable with a solution that combines the three solutions mentioned in the 'no single perfect' stack overflow article. Providing a config key to the composable that would reverse this behavior I think would also make sense, this is likely to be a special case, but I think that's OK.

This visibility check should probably be a composable itself and aspirationally, be contributed to VueUse. It is similar to [`useElementVisibility`](https://vueuse.org/core/useElementVisibility/).

**Comment by Alchez** [comment]
@agritheory, I did another pass researching on implementations, and here's what I've found:

- We could just rely on VueUse's [`useElementVisibility`](https://vueuse.org/core/useElementVisibility/) composable since [their implementation](https://github.com/vueuse/vueuse/blob/main/packages/core/useElementVisibility/index.ts) is what I've seen recommended on most forums, and should cover most expected scenarios.
- We could look into using the [asynchronous Intersection Observer API](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) instead, which seems to be made for modern situations like lazy-loading/infinite scrolling. This would also give us the benefit of reacting to changes in the viewport to apply event listeners (e.g. when a row is expanded in a tree). In addition to the docs, an [example use-case](https://stackoverflow.com/a/68961408) is also available in the same StackOverflow thread linked above.

Let me know what you think.

**Comment by agritheory** [comment]
@Alchez Let's use `useElementVisibility` until we run up against situations like the one you describe. At that point I think we'll have better understanding of our needs. Writing a wrapper that combines `useElementVisibility` and intersection observer into a common API may or may not be a great idea.

**Comment by agritheory** [comment]
First Milestone scope:

| Key(s)        | Function    |
| :------------ | :--------   |
| Enter         | `downCell`  |
| Tab           | `nextCell`  |
| Shift + Enter | `upCell`    |
| Shift+ Tab    | `prevCell`  |
| &#8592;       | `prevCell`  |
| &#8593;       | `upCell`    |
| &#8594;       | `nextCell`  |
| &#8595;       | `downCell`  |
| Home          | `firstCell`  |
| End           | `lastCell` |
| Ctrl + &#8592;   | `firstCell`  |
| Ctrl + &#8594;   | `lastCell` |

Second milestone:
| Key(s)        | Function    |
| :------------ | :--------   |
| F2         | initialize modal  |
| Alt+&#8592;       | initialize modal |
| Alt+&#8593;       | initialize modal |
| Alt+&#8594;       | initialize modal |
| Alt+&#8595;       | initialize modal |
| Ctrl + &#8593;       | first row, same column |
| Ctrl + &#8595;       | last row, same column  |

Third milestone:
| Key(s)        | Function    |
| :------------ | :--------   |
| Ctrl + C       | copy cell contents |
| Ctrl + V       | paste cell contents  |

Fourth milestone:
| Key(s)        | Function    |
| :------------ | :--------   |
| Ctrl + Shift + G | toggle focused group expand/ collapse |
| Alt+ Shift + G | toggle all group expand/ collapse |



Lasso selection is out of scope for this ticket and composable.

**Comment by Alchez** [comment]
@agritheory, I'm moving the shortcut comment into [its own task](https://github.com/agritheory/stonecrop/issues/18) to separate it from the hidden-element discussion in this issue
